pkg/build/registry/build/strategy.go

```diff
@@ -0,0 +1,121 @@
+var Decorator = func(obj runtime.Object) error {
+	build := obj.(*api.Build)
+	if build.StartTimestamp == nil {
+		build.Duration = time.Duration(0)
+	} else {
+		completionTimestamp := build.CompletionTimestamp
+		if completionTimestamp == nil {
+			dummy := util.Now()
+			completionTimestamp = &dummy
+		}
+		build.Duration = completionTimestamp.Rfc3339Copy().Time.Sub(build.StartTimestamp.Rfc3339Copy().Time)
+	}
+	return nil
+}
+
+// Strategy is the default logic that applies when creating and updating ClusterPolicy objects.
+var Strategy = strategy{kapi.Scheme}
+
+func (strategy) NamespaceScoped() bool {
+	return true
+}
+
+// AllowCreateOnUpdate is false for policies.
+func (strategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+func (strategy) GenerateName(base string) string {
+	return base
+}
+
+// PrepareForCreate clears fields that are not allowed to be set by end users on creation.
+func (strategy) PrepareForCreate(obj runtime.Object) {
+	build := obj.(*api.Build)
+
+	// TODO this is preserving the old behavior, ideally we could do this in generateName but generateName
+	// does not have access to the full build object
+	if len(build.Name) == 0 {
+		if len(build.Labels[api.BuildConfigLabel]) != 0 {
+			build.Name = fmt.Sprintf("%s-%s", build.Labels[api.BuildConfigLabel], uuid.NewUUID().String())
+		} else {
+			build.Name = uuid.NewUUID().String()
+		}
+	}
+
+	if len(build.Status) == 0 {
+		build.Status = api.BuildStatusNew
+	}
+
+}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (strategy) PrepareForUpdate(obj, old runtime.Object) {
+	_ = obj.(*api.Build)
+}
+
+// Validate validates a new policy.
+func (strategy) Validate(ctx kapi.Context, obj runtime.Object) fielderrors.ValidationErrorList {
+	return validation.ValidateBuild(obj.(*api.Build))
+}
+
+// ValidateUpdate is the default update validation for an end user.
+func (strategy) ValidateUpdate(ctx kapi.Context, obj, old runtime.Object) fielderrors.ValidationErrorList {
+	// TODO: distinguish updates from create, for now, this is preserving old implementation behavior
+	return validation.ValidateBuild(obj.(*api.Build))
+}
+
+// CheckGracefulDelete allows a build to be gracefully deleted.
+func (strategy) CheckGracefulDelete(obj runtime.Object, options *kapi.DeleteOptions) bool {
+	return false
+}
+
+// Matcher returns a generic matcher for a given label and field selector.
+func Matcher(label labels.Selector, field fields.Selector) generic.Matcher {
+	return &generic.SelectionPredicate{
+		Label: label,
+		Field: field,
+		GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+			build, ok := obj.(*api.Build)
+			if !ok {
+				return nil, nil, fmt.Errorf("not a build")
+			}
+			return labels.Set(build.ObjectMeta.Labels), SelectableFields(build), nil
+		},
+	}
+}
+
+// SelectableFields returns a label set that represents the object
+func SelectableFields(build *api.Build) fields.Set {
```

Comment:
Good point, I ran into this as well, my rush to just want to filter builds by status via field selector that led to this larger change, probably a good time to fix it up in general here.